### PR TITLE
.Net: [fix] StepwisePlanner ChatHistory token calculation refactoring

### LINKED ISSUE #2773: Chat history serialization for token counting is broken in StepwisePlanner
**Describe the bug**
When tokens are counted in **StepwisePlanner**, messages serialize incorrectly: [Code URL](https://github.com/microsoft/semantic-kernel/blob/cfebcaec6c1656f2c129ca3e40442c0ace9495e5/dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs#L447C62-L447C62)

The outcome looks like this:

> Microsoft.SemanticKernel.AI.ChatCompletion.ChatHistory+ChatMessage
Microsoft.SemanticKernel.AI.ChatCompletion.ChatHistory+ChatMessage
Microsoft.SemanticKernel.AI.ChatCompletion.ChatHistory+ChatMessage
Microsoft.SemanticKernel.AI.ChatCompletion.ChatHistory+ChatMessage

**To Reproduce**
1. Put a breakpoint in **GetChatHistoryTokens**
2. Run **StepwisePlanner**
3. See **messages** string content

**Expected behavior**
Message content should be concatenated.

**Screenshots**

**Platform**
 - OS: Windows
 - IDE: Visual Studio
 - Language: C#
 - Source: NuGet package version 0.23.230906.2-preview

**Additional context**

### PR DESCRIPTION
**Pull request by lemillermicrosoft**
Refactor chat history token calculation and clipping. Leaving the existing `ChatHistory` as-is to ensure a complete transcript is available. Extended class `ChatHistory` to support insert.

Fixes #2773 

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

File: dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs [hunk 1]
```diff
@@ -444,9 +438,15 @@ private async Task<string> GetCompletionAsync(IAIService aiService, ChatHistory
         throw new SKException("No AIService available for getting completions.");
     }
 
-    private int GetChatHistoryTokens(ChatHistory chatHistory)
+    private int GetChatHistoryTokens(ChatHistory chatHistory, string? additionalMessage = null, int skipStart = 0, int skipCount = 0)
```
Comment (SergeyMenshykh) [inline dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs:441]: nit: We are not getting tokens here, right? We are just counting them. The `CountChatHistoryTokens` method name would convey that better.

File: dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs [hunk 2]
```diff
@@ -382,34 +382,28 @@ private Task<string> GetSystemMessage(SKContext context)
...
+        var reducedChatHistory = new ChatHistory();
+        reducedChatHistory.AddRange(chatHistory.Where((m, i) => i < skipStart || i >= skipStart + skipCount));
 
-            chatHistory.Messages.RemoveAt(removalIndex);
-            tokenCount = this.GetChatHistoryTokens(chatHistory);
-            messagesRemoved++;
+        if (skipCount > 0)
+        {
+            reducedChatHistory.InsertMessage(skipStart, AuthorRole.Assistant, TrimMessage);
+            reducedChatHistory.InsertMessage(skipStart, AuthorRole.Assistant, TrimMessage);
```
Comment (SergeyMenshykh) [inline dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs:403]: Is it expected to insert the same message two times?

File: dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs (hunk 1 above)
Comment (RogerBarreto) [reply dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs:441]: Or... `GetChatHistoryTokenCount` 😊

**Review by RogerBarreto — APPROVED** [review]
nit. LGTM

File: dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs [hunk 3]
```diff
@@ -444,9 +438,15 @@ private async Task<string> GetCompletionAsync(IAIService aiService, ChatHistory
         throw new SKException("No AIService available for getting completions.");
     }
 
-    private int GetChatHistoryTokens(ChatHistory chatHistory)
+    private int GetChatHistoryTokens(ChatHistory chatHistory, string? additionalMessage = null, int skipStart = 0, int skipCount = 0)
     {
-        var messages = string.Join("\n", chatHistory.Messages);
+        var messages = string.Join("\n", chatHistory.Where((m, i) => i < skipStart || i >= skipStart + skipCount).Select(m => m.Content));
+
+        if (!string.IsNullOrEmpty(additionalMessage))
```
Comment (RogerBarreto) [inline dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs:445]: Is there any reason why the GP3Tokenizer is not being used in here.

My suggestion would also to move this method to be a ChatHistory extension.

File: dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs (hunk 2 above)
Comment (lemillermicrosoft) [reply dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs:403]: Yikes, how in the merge resolution did that happen. Thanks!

File: dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs [hunk 4]
```diff
@@ -421,7 +415,7 @@ private async Task<string> GetCompletionAsync(IAIService aiService, ChatHistory
         }
         else if (aiService is ITextCompletion textCompletion)
         {
-            var thoughtProcess = string.Join("\n", chatHistory.Messages.Select(m => m.Content));
+            var thoughtProcess = string.Join("\n", chatHistory.Select(m => m.Content));
```
[the remark this answers is not in the document]
Comment (lemillermicrosoft) [reply dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs:418]: There is no `public static String Join(char separator, IEnumerable<String> values);`

**Pull request merged**